### PR TITLE
Feature/api basic auth

### PR DESCRIPTION
**Pull request by OwenTuz**
#10615 - Add basic auth to REST API

```
Adds @requires_auth decorator to API, which checks basic auth against a SHA512-hashed password
Adds CLI options for 'username' and 'password'
Adds YAML configuration option for API (using CLI option "-c /path/to/file.yaml")
Updates API tests, Docker API script and documentation to reflect the change
```

Note that this breaks backwards compatibility by forcing a requirement for authentication on all views (ie users must specify a password hash either on the CLI or in a configuration file - there's no default). It also requires users running the Docker script to specify a configuration file containing at minimum their hashed password.

I'm happy to refactor that if we would prefer to make basic authentication optional and retain backward compatibility.

### CODE REVIEW

File: doc/sphinx/api/index.rst [hunk 1]
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
...
+asks users to create and specify a configuration file.
+
+Authentication
+--------------
+
+Linux or Mac users can generate a SHA512 hash from a plaintext password by
+running:
+
+.. code-block:: none
+    $ echo -n "secret" | sha512sum
```
Comment (andresriancho) [inline doc/sphinx/api/index.rst]: Add empty line after the `code-block` directive

File: doc/sphinx/api/index.rst [hunk 2]
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
...
+    bd2b1aaf7ef4f09be9f52ce2d8d599674d81aa9d6a4421696dc4d93dd0619d682ce56b4d64a9ef097761ced99e0f67265b5f76085e5b0ee7ca4696b2ad6fe2b2  -
+
+If the above hash was specified using the "password" option, users would be able
```
Comment (andresriancho) [inline doc/sphinx/api/index.rst]: Replace was with is

File: doc/sphinx/api/index.rst [hunk 3]
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
...
+    bd2b1aaf7ef4f09be9f52ce2d8d599674d81aa9d6a4421696dc4d93dd0619d682ce56b4d64a9ef097761ced99e0f67265b5f76085e5b0ee7ca4696b2ad6fe2b2  -
+
+If the above hash was specified using the "password" option, users would be able 
+to authenticate using HTTP basic auth with the default username 'admin' and 
+password 'secret'.
+
+For example, using the 'curl' command:
+
+.. code-block:: none
+    $ curl -u admin:secret http://127.0.0.1:5000
```
Comment (andresriancho) [inline doc/sphinx/api/index.rst]: Add empty line after `code-block`

File: doc/sphinx/api/index.rst [hunk 4]
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
...
+    {
+      "docs": "http://docs.w3af.org/en/latest/api/index.html" 
+    }
+
+Although this provides some security, traffic passing to and from the API is not
```
Comment (andresriancho) [inline doc/sphinx/api/index.rst]: > Please note that even with basic authentication, traffic passing to and from the API is not

File: doc/sphinx/api/index.rst [hunk 5]
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
...
+sniffed by an attacker with "man-in-the-middle" capabilities.
+
+If running the API on a publicly available IP address we recommend taking
+additional precautions including running it behind an SSL proxy server (such as 
+Pound, or Apache with mod_proxy enabled).
+
+Config file format
+------------------
+
+Using a configuration file is completely optional unless you're starting the
```
Comment (andresriancho) [inline doc/sphinx/api/index.rst]: Remove "completely"

File: extras/docker/scripts/w3af_api_docker [hunk 6]
```diff
@@ -15,6 +15,9 @@ def parse_args():
                                             ' is stable, try unstable to get'
                                             ' the latest w3af features.',
                         required=False)
+    parser.add_argument('-c', '--config-file',
+                        help='[Required] Path to API config file in YAML format.',
```
Comment (andresriancho) [inline extras/docker/scripts/w3af_api_docker]: Remove the `[Required]`, argparse will show the required arguments in a different way to the user so this text is redundant

File: w3af/core/ui/api/main.py [hunk 7]
```diff
@@ -51,17 +52,47 @@ def parse_arguments():
...
                         help='Specify address where the REST API will listen'
                              ' for HTTP requests. If not specified 127.0.0.1:'
                              '5000 will be used.',
-                        default='127.0.0.1:5000',
-                        nargs='?',
-                        type=parse_host_port)
+                        default=False,
+                        nargs='?')
+
+    parser.add_argument('-c',
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: I believe that argparse provides a "type" for input files, if you specify that argparse will make sure the file exists, is readable, etc.

File: w3af/core/ui/api/main.py [hunk 8]
```diff
@@ -86,16 +121,55 @@ def main():
...
+      try:
+        with open(args.config_file) as f:
+          yaml_conf = yaml.safe_load(f)
+      except:
+        print('Error loading config file %s. Please check it exists and is'
+              ' a valid YAML file.' % args.config_file)
+        return 1
 
-    if args.host not in ('127.0.0.1', 'localhost'):
-        print('The REST API does not provide authentication and might expose'
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: Maybe replace this message with a warning about SSL? (like in the docs)

File: w3af/core/ui/api/main.py [hunk 9]
```diff
@@ -70,10 +101,14 @@ def parse_arguments():
 
     args = parser.parse_args()
 
-    host_port = getattr(args, 'host:port')
-    args.host = host_port[0]
-    args.port = host_port[1]
-
+    try:
+      args.host, args.port = getattr(args,'host:port').split(':')
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: Please use 4-spaces for indent (this applies to all the code)

File: w3af/core/ui/api/main.py [hunk 10]
```diff
@@ -86,16 +121,55 @@ def main():
...
+    for k in defaults:
+      if not k in app.config:
+        app.config[k] = defaults[k]
+
+    app.config['HOST'], app.config['PORT'] = parse_host_port(app.config['HOST'],
+                                                             app.config['PORT'])
+
+    if (app.config['HOST'] == '127.0.0.1' or
+        app.config['HOST'] == 'localhost'):
+      print('CAUTION! Traffic to this API is not encrypted and could be sniffed.'
```
Comment (andresriancho) [inline w3af/core/ui/api/main.py]: Ah! You did add this :+1:

**Comment by andresriancho** [comment]
I like your PR and do believe that we should let the user start the API without authentication (default). When starting without auth we should print a warning with the two issues (no auth, no ssl).

**Comment by OwenTuz** [comment]
Thanks for taking a look over it! I've updated this based on the above.

- Various code and documentation formatting fixes
- Authentication is now optional and only applied if a password is specified on the CLI or in a config file
- Unit tests still use authentication
- Fixed a daft bug where I had specified '==' instead of '!=' (...)
- Switched to using argparse.FileType to test file existence.
- Tightened up type checking when copying key/value pairs into Flask's app.config
- Made 'debug' and 'config-file' options mutually exclusive arguments in the Docker script (since 'debug' runs bash and bypasses the API command line entirely)

**Comment by andresriancho** [comment]
It all looks good now, anything else that's pending? If it looks good to you, I'll git pull your code, test it live and merge if all works

**Comment by OwenTuz** [comment]
I think that's the lot - please do. You know where to find me when you spot the bugs.

**Comment by andresriancho** [comment]
Ok - will review this tomorrow

**Comment by andresriancho** [comment]
Take a look at https://github.com/andresriancho/w3af/commit/4e1a4aa482b4f4dc62128a20ba6eb7089d04c855 and https://github.com/andresriancho/w3af/commit/b18018f31f5f96adc10282fcee9099a1169ebb59 for a couple of improvements on your already great work :+1:

**Comment by andresriancho** [comment]
Tests are running at https://circleci.com/gh/andresriancho/w3af/2140 , after they PASS I'll merge into develop

**Comment by andresriancho** [comment]
Build success, merging! Thanks for the contribution, keep them coming, some suggestions:
 * https://github.com/andresriancho/w3af/issues/10831
 * https://github.com/andresriancho/w3af/issues/10616

**Comment by OwenTuz** [comment]
Thanks :) Will take a look.

**Pull request merged**